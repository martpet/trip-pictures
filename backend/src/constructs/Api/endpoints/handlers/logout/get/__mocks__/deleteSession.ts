export const deleteSession = vi
  .fn()
  .mockName('deleteSession')
  .mockResolvedValue('dummyRefreshToken');
