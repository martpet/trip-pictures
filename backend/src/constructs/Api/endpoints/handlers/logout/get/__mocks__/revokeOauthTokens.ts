export const revokeOauthTokens = vi
  .fn()
  .mockName('revokeOauthTokens')
  .mockResolvedValue(undefined);
