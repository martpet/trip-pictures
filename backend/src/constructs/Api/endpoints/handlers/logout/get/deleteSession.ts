import {
  DeleteCommand,
  DynamoDBClient,
  DynamoDBDocumentClient,
  SessionsTableItem,
  sessionsTableOptions,
} from 'lambda-layer';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

export const deleteSession = async (sessionId: string) => {
  const deleteCommand = new DeleteCommand({
    TableName: sessionsTableOptions.tableName,
    Key: { [sessionsTableOptions.partitionKey.name]: sessionId },
    ReturnValues: 'ALL_OLD',
  });

  const { Attributes } = await ddbDocClient.send(deleteCommand);
  const sessionsItem = Attributes as SessionsTableItem | undefined;

  if (!sessionsItem) {
    throw new Error(`could not find session with id "${sessionId}"`);
  }

  return sessionsItem.refreshToken;
};
