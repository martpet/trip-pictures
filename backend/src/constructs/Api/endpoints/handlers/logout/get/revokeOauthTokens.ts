import { authPaths } from 'lambda-layer';
import fetch from 'node-fetch';

type RevokeTokensProps = {
  authDomain: string;
  refreshToken: string;
  clientId: string;
};

export const revokeOauthTokens = async ({
  authDomain,
  refreshToken,
  clientId,
}: RevokeTokensProps) => {
  const response = await fetch(`https://${authDomain}${authPaths.revoke}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      token: refreshToken,
      client_id: clientId,
    }),
  });

  if (!response.ok) {
    const responseText = await response.text();
    throw new Error(responseText);
  }
};
