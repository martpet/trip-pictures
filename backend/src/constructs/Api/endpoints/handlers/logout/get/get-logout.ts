import {
  APIGatewayProxyHandlerV2,
  authPaths,
  cookie,
  cookieName,
  EnvVars,
  errorResponse,
  localhostUrl,
  parseEventCookies,
  RouteHeaders,
  StatusCodes,
} from 'lambda-layer';

import { deleteSession } from './deleteSession';
import { revokeOauthTokens } from './revokeOauthTokens';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const { envName } = globalLambdaProps;
  const { authDomain, clientId, logoutCallbackUrl, logoutCallbackLocalhostUrl } =
    process.env as EnvVars<'/logout', 'GET'>;
  const { referer } = event.headers as RouteHeaders<'/logout'>;
  const { sessionId } = parseEventCookies<'/logout'>(event);
  const isFromLocalhost = envName === 'personal' && referer?.startsWith(localhostUrl);
  const cognitoLogoutUrl = new URL(`https://${authDomain}${authPaths.logout}`);
  const blankSessionCookie = cookie.serialize(cookieName('sessionId'), '', {
    expires: new Date(),
  });

  if (!authDomain || !clientId || !logoutCallbackUrl || !logoutCallbackLocalhostUrl) {
    return errorResponse('WKSoIIT1Ds');
  }

  if (envName === 'personal' && !referer) {
    return errorResponse('GkjK-mpVvL');
  }

  if (sessionId) {
    const refreshToken = await deleteSession(sessionId);
    await revokeOauthTokens({ authDomain, clientId, refreshToken });
  }

  const cognitoLogoutUrlParams = new URLSearchParams({
    client_id: clientId,
    logout_uri: isFromLocalhost ? logoutCallbackLocalhostUrl : logoutCallbackUrl,
  });

  cognitoLogoutUrl.search = cognitoLogoutUrlParams.toString();

  return {
    statusCode: StatusCodes.MOVED_TEMPORARILY,
    headers: { location: cognitoLogoutUrl.href },
    cookies: [blankSessionCookie],
  };
};
