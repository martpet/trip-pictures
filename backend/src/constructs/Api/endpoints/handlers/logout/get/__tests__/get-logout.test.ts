import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import {
  itCalls,
  itHasEnvVars,
  itResolves,
  itResolvesWithError,
  parseEventCookies,
} from '~/constructs/Api/utils';

import { deleteSession } from '../deleteSession';
import { handler } from '../get-logout';
import { revokeOauthTokens } from '../revokeOauthTokens';

vi.mock('../deleteSession');
vi.mock('../revokeOauthTokens');
vi.mock('~/constructs/Api/utils/parseEventCookies');
vi.mock('~/constructs/Api/utils/errorResponse');

process.env.authDomain = 'dummyAuthDomain';
process.env.clientId = 'dummyClientId';
process.env.logoutCallbackUrl = 'dummyLogoutCallbackUrl';
process.env.logoutCallbackLocalhostUrl = 'dummyLocalhostLogoutCallbackUrl';

const args = [
  {
    cookies: 'dummyCookies',
    headers: {
      referer: 'dummyRefererHeader',
    },
  },
] as unknown as Parameters<APIGatewayProxyHandlerV2>;

describe('"get-logout" handler', () => {
  itHasEnvVars(
    ['authDomain', 'clientId', 'logoutCallbackUrl', 'logoutCallbackLocalhostUrl'],
    handler,
    args
  );
  itCalls(parseEventCookies, handler, args);
  itResolves(handler, args);

  describe('when "parseEventCookies" returns a "sessionId"', () => {
    beforeEach(() => {
      vi.mocked(parseEventCookies).mockReturnValueOnce({ sessionId: 'dummySessionId' });
    });
    itCalls(deleteSession, handler, args);
    itCalls(revokeOauthTokens, handler, args);
  });

  describe('when "globalLambdaProps.envName" is "personal"', () => {
    const initialEnvName = globalLambdaProps.envName;
    beforeAll(() => {
      globalLambdaProps.envName = 'personal';
    });
    afterAll(() => {
      globalLambdaProps.envName = initialEnvName;
    });
    describe('when "referer" header is missing', () => {
      const argsClone = structuredClone(args);
      delete argsClone[0].headers.referer;
      itResolvesWithError(handler, argsClone);
    });
    describe('when "referer" is from localhost', () => {
      const argsClone = structuredClone(args);
      argsClone[0].headers.referer = 'http://localhost:3000/dummyPath';
      itResolves(handler, argsClone);
    });
  });
});
