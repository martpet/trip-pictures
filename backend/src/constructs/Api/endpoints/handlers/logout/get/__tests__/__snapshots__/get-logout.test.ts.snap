// Vitest Snapshot v1

exports[`"get-logout" handler > calls "parseEventCookies" with correct args 1`] = `
[
  [
    {
      "cookies": "dummyCookies",
      "headers": {
        "referer": "dummyRefererHeader",
      },
    },
  ],
]
`;

exports[`"get-logout" handler > resolves with a correct value 1`] = `
{
  "cookies": [
    "sessionId=; Expires=Sat, 01 Jan 2022 00:00:00 GMT",
  ],
  "headers": {
    "location": "https://dummyauthdomain/logout?client_id=dummyClientId&logout_uri=dummyLogoutCallbackUrl",
  },
  "statusCode": 302,
}
`;

exports[`"get-logout" handler > when "authDomain" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "WKSoIIT1Ds",
  ],
]
`;

exports[`"get-logout" handler > when "authDomain" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-logout" handler > when "clientId" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "WKSoIIT1Ds",
  ],
]
`;

exports[`"get-logout" handler > when "clientId" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-logout" handler > when "globalLambdaProps.envName" is "personal" > when "referer" header is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "GkjK-mpVvL",
  ],
]
`;

exports[`"get-logout" handler > when "globalLambdaProps.envName" is "personal" > when "referer" header is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-logout" handler > when "globalLambdaProps.envName" is "personal" > when "referer" is from localhost > resolves with a correct value 1`] = `
{
  "cookies": [
    "sessionId=; Expires=Sat, 01 Jan 2022 00:00:00 GMT",
  ],
  "headers": {
    "location": "https://dummyauthdomain/logout?client_id=dummyClientId&logout_uri=dummyLocalhostLogoutCallbackUrl",
  },
  "statusCode": 302,
}
`;

exports[`"get-logout" handler > when "logoutCallbackLocalhostUrl" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "WKSoIIT1Ds",
  ],
]
`;

exports[`"get-logout" handler > when "logoutCallbackLocalhostUrl" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-logout" handler > when "logoutCallbackUrl" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "WKSoIIT1Ds",
  ],
]
`;

exports[`"get-logout" handler > when "logoutCallbackUrl" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-logout" handler > when "parseEventCookies" returns a "sessionId" > calls "deleteSession" with correct args 1`] = `
[
  [
    "dummySessionId",
  ],
]
`;

exports[`"get-logout" handler > when "parseEventCookies" returns a "sessionId" > calls "revokeOauthTokens" with correct args 1`] = `
[
  [
    {
      "authDomain": "dummyAuthDomain",
      "clientId": "dummyClientId",
      "refreshToken": "dummyRefreshToken",
    },
  ],
]
`;
